### PR TITLE
add ParameterList type

### PR DESCRIPTION
**Pull request by WalterBright**
`class Parameter` is an inside-out abstraction, whereby an element of an array attempts to encapsulate the array. Adding `ParameterList` to encapsulate the array moves towards fixing this. Eventually, I want to move it towards supporting range operations, and include the `varargs` as part of it. Initially, range operations `.length` and `[i]` are supported.

In the meantime, it still needs to be compatible with the C++ interface, and with a couple of `alias` declarations the rest of dmd can be incrementally moved toward using ParameterList.

### CODE REVIEW

File: src/dmd/mtype.d [hunk 1]
```diff
@@ -6183,6 +6183,27 @@ extern (C++) final class TypeNull : Type
...
+struct ParameterList
+{
+    Parameters* parameters;
+
+    size_t length()
+    {
+        return Parameter.dim(parameters);
+    }
+
+    Parameter opIndex(size_t i)
```
Comment (thewilsonator) [inline src/dmd/mtype.d]: `ref opIndex`?

Comment (WalterBright) [reply src/dmd/mtype.d]: A `Parameter` is a class, and is already by ref.

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Buildkite fails with 
```
mv ./res dmd
--
  | mv: cannot move './res' to 'dmd/res': Directory not empty
```
which is surely unrelated.

**Comment by thewilsonator** [comment]
Gah, need to update the C++ headers as well.

File: src/dmd/mtype.d [hunk 2]
```diff
@@ -625,9 +625,7 @@ extern (C++) abstract class Type : RootObject
         }
         else if (t1.parameters != t2.parameters)
         {
-            size_t dim1 = !t1.parameters ? 0 : t1.parameters.dim;
-            size_t dim2 = !t2.parameters ? 0 : t2.parameters.dim;
-            if (dim1 || dim2)
+            if (t1.parameterList.length || t1.parameterList.length)
```
Comment (nucz) [inline src/dmd/mtype.d]: missed t2

**Pull request merged**